Allow migrated domains to redirect form submissions to new host

kaapstorm opened the pull request:
## Product Description

This change allows domains that have been migrated to third-party-managed hosts to redirect form submissions, syncs, case claim requests etc. to their new host.

This avoids problems when deployed CommCare apps are not using a DNS alias prior to the migration.

## Technical Summary

Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field. The redirect URL is set using a management command. While this has the disadvantage of needing to be done by a developer, it has the advantages that a developer can ensure the URL is legitimate, and that the URL can be set after the domain has been locked by the data migration feature flag.

Jira: [SC-3628](https://dimagi.atlassian.net/browse/SC-3628)

Documentation: [Redirect on Data Migration](https://docs.google.com/document/d/1mQxRmboW3CuOkmklUjf5jypgJcv9VX_e26TEiFh1CyI/edit)

FYI @ctsims 

## Feature Flag

DATA_MIGRATION

## Safety Assurance

### Safety story

* HQ behavior is verified with unit tests
* Interaction with mobile to be verified by QA

### Automated test coverage

Tests verify the management command, and the view decorator that protects all URLs for submitting and fetching form and case data.

### QA Plan

QA ticket: [QA-6406](https://dimagi.atlassian.net/browse/QA-6406)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SC-3628]: https://dimagi.atlassian.net/browse/SC-3628?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-6406]: https://dimagi.atlassian.net/browse/QA-6406?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Domain.redirect_url and management command

kaapstorm committed:
Tests

Charl1996 reviewed (approved):
Looks good to me. 

Just a note: I'm pro streamlining this workflow in the future (not now, since this is a rather high priority ticket), thus having some place in HQ that this could be done, so not having to involve a developer...

mkangia commented on corehq/apps/domain/decorators.py:
```diff
@@ -668,6 +679,10 @@ def _inner(request, *args, **kwargs):
         if DATA_MIGRATION.enabled(domain):
+            domain_obj = Domain.get_by_name(domain)
+            if domain_obj.redirect_url:
+                return HttpResponsePermanentRedirect(domain_obj.redirect_url)
```
just checking if we want to consider making a redirect and not a permanent redirect?
I don't know if permanent redirect is an absolute requirement.

I say so because if the migration isn't successful and the project wants to start using SaaS environment again, they won't have a way to do that then? Just good to confirm/communicate that

kaapstorm replied:
I agree that we should communicate this very clearly.

They should not use this before the migration is completed, and confirmed to have been successful.

If they enable this before then, then they _will lose data_, and the point of this feature is to avoid them losing data.

millerdev replied:
But why not make this a non-permanent redirect?

kaapstorm replied:
I chose a permanent redirect because once data has been migrated to the new environment, syncing with the old environment will always result in data loss.

We could use a non-permanent redirect (HTTP 302: Found) but I'm not sure that the meaning of the code is one that we want to convey. The device should, from the moment the data migration has completed successfully, ever be syncing with the old environment again. ... Or am I missing a scenario where that is not true?

kaapstorm replied:
Offlined about this. I've switched to a temporary redirect, because people need to be able to change their minds. https://github.com/dimagi/commcare-hq/pull/34461/commits/69488f09934c37513b7ec8727921ff1b4a82c959

mkangia reviewed (approved):
Great tests.

Just one query that could really be non blocking if it has already been thought about or can't be done.

kaapstorm committed:
Lint

kaapstorm committed:
Smart redirect to support all endpoints

esoergel commented on corehq/apps/domain/models.py:
```diff
@@ -449,6 +449,9 @@ class Domain(QuickCachedDocumentMixin, BlobMixin, Document, SnapshotMixin):
     orphan_case_alerts_warning = BooleanProperty(default=False)
 
+    # For domains that have been migrated to a different environment
+    redirect_url = StringProperty()
```
> Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field.

No objections here, but I do have some passing idle thoughts.  I have also wondered how best to deal with these random one-off domain settings.  The obvious one would be a `DomainSettings` (sql) model that's got a bunch of columns on it.  There are few enough domains that we don't really need to worry about the scale of data stored.

Another possibility that occurs to me, especially for sparsely populated fields is to use an entity-attribute-value model that doesn't require a migration for new fields.  Something like:
```python
class DomainSetting(models.Model):
    domain = models.CharField(db_index=True)
    slug = models.CharField(db_index=True)
    value = models.HmmActuallyMaybeThisIsHarderThanIThought()
```
Only thing I don't like about that is you'd have to use one fixed schema for `value`, which might be a string or a boolean or something else, depending on the setting.  Okay maybe a simple `DomainSettings` model _is_ best.

kaapstorm replied:
675f785d3c4

mkangia replied:
if we are creating a new model, how about exposing it to django admin?

kaapstorm replied:
43efa33b1b9

kaapstorm committed:
Move `GetOrNoneManager` to corehq.util.models

kaapstorm committed:
Move new Domain property into DomainSettings

kaapstorm committed:
DomainSettings dump/reload and deletion

kaapstorm committed:
Move validation to the model

kaapstorm committed:
Add to Django Admin

mkangia reviewed (commented):
Cool, thanks for the follow up @kaapstorm 

Looks good. Will wait for QA for final approval.

mkangia commented:
> ^^^ No. The migrations are fast (they just create a table), but they refer to a field type that is defined in the code, and so they need to be applied at the same time.

Curious about this. I believe that check on the PR is to ensure that the migration won't impact the currently running code on the environment. A scenario where till be a concern is if you were to remove a column, that will be referenced on the environment till the code is updated, which would be bad. 
So that check should be okay on this PR since you are just adding a new table which will only be referenced post code update.
Right?

millerdev commented on corehq/apps/domain/models.py:
```diff
@@ -879,6 +889,47 @@ def get_daily_outbound_sms_limit(self):
+        super().validate(value, model_instance)
+
+        if not value.startswith('https'):
+            raise ValidationError(f'{value} is not a secure URL.')
```
Will this block setting the value to an empty string? Or are validators not executed on empty values?

Also, this seems like it's enforcing a policy decision that may not match real world circumstances. If someone moved their domain to self-hosted and decided not to put an SSL certificate on it, are we not going to redirect to that domain? I think this should be a strong suggestion (warning if violated, probably implemented in UI JS) rather than a hard requirement.

millerdev commented on migrations.lock:
```diff
@@ -390,6 +390,8 @@ domain
  0013_accountconfirmationsettings_squashed_0016_alter_smsaccountconfirmationsettings_project_name (4 squashed migrations)
  0014_appreleasemodesetting
+ 0015_domainsettings
+ 0016_auto_20240419_2223
```
Can these migrations be squashed? If not, would be nice if the second one had a nicer name.

millerdev reviewed (commented):
The description says:

> Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field.

I'm confused. The current solution in this PR looks like what the description said it was not. Notably, it does not conform with...

Ethan suggested:

> Another possibility that occurs to me ... is to use an entity-attribute-value model that doesn't require a migration for new fields.

I like this suggestion. We could use a `JSONField` to solve the `HmmActuallyMaybeThisIsHarderThanIThought` field type problem. Although that still has a disadvantage of not easily being able to store dates. We could work around that by making an `ExoticValueJsonField` that would auto-convert certain values to rich types (example: `{"__type__": "date", "value": "2024-04-01"}` -> `date(2024, 4, 1)`). Luckily we don't need to solve that problem immediately, so a plain `JSONField` would work for now.

kaapstorm committed:
Revert all DjangoSettings model changes

millerdev commented:
I see one [unanswered question](https://github.com/dimagi/commcare-hq/pull/34461#discussion_r1574788808). Otherwise, I think this is good to go.

kaapstorm committed:
Use temporary redirect

kaapstorm committed:
"An unexpected issue occurred ..." Trying 302.

kaapstorm committed:
Mobile needs 301

Also allow redirects when the Data Migration feature flag has been
disabled, to allow Mobile to log in.

*The feature flag still needs to be enabled when the redirect is set!*

kaapstorm committed:
Back to 302

millerdev reviewed (approved):
I'm curious about the things that motivated each of the last three commits. Can you share those details?

kaapstorm commented:
The mobile team has coalesced on 302.

I've also tweaked the decorator to work when the feature flag is disabled, to allow mobile users to log in, but the feature flag still needs to be enabled to set the redirect URL. So the workflow for data migrations is:
1. Set the Data Migration feature flag
2. Do the actual migration
3. Set the redirect URL to the new environment so that devices can't sync with the old environment.
4. Disable the feature flag so that mobile users can log in.
